app/src/lib/equality.ts

```diff
@@ -77,3 +77,33 @@ export function arrayEquals<T>(x: ReadonlyArray<T>, y: ReadonlyArray<T>) {
 
   return true
 }
+
+/**
+ * Compares two maps for key reference equality.
+ *
+ * Two maps are considered equal if all their keys coincide, if they're
+ * both empty or if they're the same object.
+ */
+export function mapKeysEqual<T>(x: Map<T, unknown>, y: Map<T, unknown>) {
+  if (x === y) {
+    return true
+  }
+
+  if (x.size !== y.size) {
+    return false
+  }
+
+  for (const key of x.keys()) {
+    if (!y.has(key)) {
+      return false
+    }
+  }
+
+  for (const key of y.keys()) {
+    if (!x.has(key)) {
+      return false
+    }
+  }
```

Comment:
I doubt it has a practical performance impact but isn't this redundant given that we short-circuit if the size differs?  Shouldn't it be enough to just check that all the keys of `x` exists in `y`?

```suggestion

  for (const key of y.keys()) {
    if (!x.has(key)) {
      return false
    }
  }
```

Reply:
Oh yes you're right!